pytorch_lightning/trainer/_progress.py

```diff
@@ -0,0 +1,113 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+
+
+@dataclass
+class LoopProgress:
+    """Basic dataclass to track loop progress during execution.
+
+    The general structure is to count totals across epochs as well as intra-epoch counts.
+    These counters are local to a trainer rank.
+    By default, they are not by default globally synced across all ranks.
+
+    Args:
+        total_epochs_finished: Number of passes through the dataset.
+        total_batches_read: Number of batches loaded through the dataloader.
+        total_batches_started: Number of batches started processing.
+            This is intended to be incremented after `on_train_batch_start` completes.
+        total_batches_processed: Number of batches processed.
+            This is intended to be incremented after `training_step` runs.
+        total_batches_finished: Number of batches finished.
+            This is intended to be incremented after `on_train_batch_end` runs.
+
+        batches_read_this_epoch: Number of batches loaded through the dataloader within the current epoch.
+        batches_started_this_epoch: Number of batches started processing within the current epoch.
+            This is intended to be incremented after `on_train_batch_start` completes.
+        batches_processed_this_epoch: Number of batches processed within the current epoch.
+            This is intended to be incremented after `training_step` runs.
+        batches_finished_this_epoch: Number of batches finished within the current epoch.
+            This is intended to be incremented after `on_train_batch_end` runs.
+    """
+    total_epochs_finished: int = 0
+
+    total_batches_read: int = 0
+    total_batches_started: int = 0
+    total_batches_processed: int = 0
+    total_batches_finished: int = 0
```

Comment:
good question. i think these should be reset whenever `fit`/`test`/`validate`/`predict` starts. otherwise when this is used for checkpointing, we'll save the wrong data inside the checkpoint dict. this should be part of the trainer state when resuming from checkpoint

i dont know how useful having a global one is and i'm paranoid about leftover state in the trainer leading to difficult bugs later on

Reply:
Okay, we can hold off the global ones for now.

> be reset whenever fit/test/validate/predict starts

> this should be part of the trainer state when resuming from checkpoint

What if this state is always part of the model and the loops just update it but not reset it?

This way `resume_from_ckpt`/`load_from_ckpt` just reloads this state to the model and the trainer does not have to reset it at all.

Reply:
could you describe why this state is part of the model?